scheme/sevsnp/evidence_handler.go

```diff
@@ -0,0 +1,530 @@
+	case ReportSigningKeyVcek:
+		if len(certChain.GetVcekCert()) == 0 {
+			return ErrNoVCEK
+		}
+		if vcek, err = readCert(certChain.GetVcekCert()); err != nil {
+			return fmt.Errorf("can't read VCEK to validate cert chain: %w", err)
+		}
+		attestation.CertificateChain = &sevsnp.CertificateChain{VcekCert: vcek, AskCert: ask, ArkCert: ark}
+	default:
+		return ErrBadSigningKey
+	}
+
+	return verify.SnpAttestation(&attestation, &options)
+}
+
+// ValidateEvidenceIntegrity confirms the integrity of evidence by doing the following:
+//   - verifies that the TA in the evidence matches the provisioned TA
+//   - confirms the integrity of the certificate chain
+//   - validates the integrity of evidence by checking its signature
+func (o EvidenceHandler) ValidateEvidenceIntegrity(
+	token *proto.AttestationToken,
+	trustAnchors []string,
+	_ []string,
+) error {
+	var (
+		tsm            *tokens.TSMReport
+		provisionedArk *comid.CryptoKey
+		certChain      *sevsnp.CertificateChain
+		err            error
+	)
+
+	if tsm, err = parseAttestationToken(token); err != nil {
+		return err
+	}
+
+	if provisionedArk, err = extractProvisionedTA(trustAnchors); err != nil {
+		return err
+	}
+
+	if certChain, err = parseCertificateChainFromEvidence(tsm); err != nil {
+		return err
+	}
+
+	if err := validateCertificateChain(certChain); err != nil {
+		return err
+	}
+
+	if err := validateTA(certChain, provisionedArk); err != nil {
+		return err
+	}
+
+	return validateReportIntegrity(tsm, certChain)
+}
+
+// refvalToComidTriple converts extracted reference values to CoMID value triple
+func refvalToComidTriple(endorsementsStrings []string) (*comid.ValueTriple, error) {
+	var (
+		refValEndorsement *handler.Endorsement
+		rv                comid.ValueTriple
+	)
+
+	for i, e := range endorsementsStrings {
+		var endorsement handler.Endorsement
+
+		if err := json.Unmarshal([]byte(e), &endorsement); err != nil {
+			return nil, fmt.Errorf("could not decode endorsement at index %d: %w", i, err)
+		}
+
+		if endorsement.Type == handler.EndorsementType_REFERENCE_VALUE {
+			refValEndorsement = &endorsement
+			break
+		}
+	}
+
+	if refValEndorsement == nil {
+		return nil, ErrNoProvisionedRV
+	}
+
+	err := json.Unmarshal(refValEndorsement.Attributes, &rv)
+	if err != nil {
+		return nil, err
+	}
+
+	return &rv, nil
+}
+
+// evidenceToComidTriple converts claim set to CoMID value triple
+func evidenceToComidTriple(ec *proto.EvidenceContext) (*comid.ValueTriple, error) {
+	evCorimJson, err := json.Marshal(ec.Evidence.AsMap())
+	if err != nil {
+		return nil, err
+	}
+
+	evComid, err := comidFromJson(evCorimJson)
+	if err != nil {
+		return nil, err
+	}
+
+	return &evComid.Triples.ReferenceValues.Values[0], nil
+}
+
+// compareMeasurements checks if two given comid.Measurement variables are the same.
+func compareMeasurements(refM comid.Measurement, evM comid.Measurement) bool {
+	// RawValue comparison
+	if refM.Val.RawValue != nil {
```

Comment:
Could you please help me understand the matching logic?  I would have expected Evidence to drive the process.  In other words, if evidence reports a raw value, digest, or SVN, we should attempt to match it against the reference values.  However, the logic is inverted here, which is confusing.

Reply:
Please go one step above the stack to `AppraiseEvidence` method and check `claimsLoop`.

The principle here is evidence must have all the claims in the Reference Value. Evidence can have extra claims, but it must _**at the least**_ have all the claims in the RV.

We ignore the extra claims not present in RV. Does that sound good?